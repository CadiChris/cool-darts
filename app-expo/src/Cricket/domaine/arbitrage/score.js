export function scoreVierge(joueur) {
  return {
    joueur,
    penalite: 0,
    cible: {
      15: { touches: 0, ferme: false },
      16: { touches: 0, ferme: false },
      17: { touches: 0, ferme: false },
      18: { touches: 0, ferme: false },
      19: { touches: 0, ferme: false },
      20: { touches: 0, ferme: false },
      25: { touches: 0, ferme: false },
    },
  };
}

const LIMITE = 3;

export function toucher(chiffre, touches, score) {
  const touchesActuelles = score.cible[chiffre].touches;
  const nouvellesTouches = Math.min(touchesActuelles + touches, LIMITE);

  return {
    ...score,
    cible: {
      ...score.cible,
      [chiffre]: {
        touches: nouvellesTouches,
        ferme: nouvellesTouches >= LIMITE,
      },
    },
  };
}

export function penaliser(pointsDePenalite, score) {
  return {
    ...score,
    penalite: score.penalite + pointsDePenalite,
  };
}

export function chiffreEstFerme(chiffre, score) {
  return score.cible[chiffre].ferme;
}
