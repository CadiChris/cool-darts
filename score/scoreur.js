"use strict";
module.exports = {

	scorer: function(tableauDesScores, lancer){ 
		let penalite = penaliteDuLancer(tableauDesScores, lancer);
		
		return tableauDesScores.map(s => {
			return scorer(s, lancer, penalite)
		});
	}
}

const LIMITE = 3;

function penaliteDuLancer(scores, lancer) {
	let scoreDuLanceur = trouverLeScoreDuLanceur(scores, lancer.lanceur);
	return calculerLaPenalite(scoreDuLanceur.cible, lancer);
}

function trouverLeScoreDuLanceur(scores, lanceur) {
	return scores.filter(s => { return s.joueur === lanceur })[0];
}

function calculerLaPenalite(cibleDuLanceur, lancer) {
	let touches = cibleDuLanceur[lancer.chiffre].touches;
	let leChiffreVaSeFermer = touches + lancer.touches >= LIMITE;
	let surplus = (touches + lancer.touches) - LIMITE;
	
	return leChiffreVaSeFermer ? surplus * lancer.chiffre : 0;
}

function scorer(score, lancer, penalite) {
	return score.joueur === lancer.lanceur
		? scorerLeLanceur(score, lancer.chiffre, lancer.touches)
		: scorerUnAdversaire(score, lancer.chiffre, penalite);
}

function scorerLeLanceur(scoreActuel, chiffre, touches) {
	let cibleDuLanceur = scoreActuel.cible;
	
	let nouvellesTouches = touchesApresLeLancer(scoreActuel.cible, chiffre, touches);
	let nouvelleCible = toucherLaCible(cibleDuLanceur, chiffre, nouvellesTouches);
	
	let nouveauScore = Object.assign({}, scoreActuel, {
		cible: nouvelleCible
	});
	
	return nouveauScore;
}

function touchesApresLeLancer(cible, chiffre, touchesDuLancer) {
	let ancienCompte = cible[chiffre].touches;
	let nouveauCompte = Math.min(ancienCompte + touchesDuLancer, LIMITE);
	return nouveauCompte;
}

function toucherLaCible(cible, chiffreTouche, touches) {
	return Object.assign({}, cible, {
		[chiffreTouche] : {
			touches: touches,
			ferme: touches >= LIMITE
		}
	});
}

function scorerUnAdversaire(score, chiffre, nombreDePoints) {
	let nouveauxPoints = score.cible[chiffre].ferme
			? score.points
			: score.points + nombreDePoints;
	
	return Object.assign({}, score, {
		points: nouveauxPoints
	});
}
