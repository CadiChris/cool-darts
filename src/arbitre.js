"use strict";
module.exports = {

	scorer: function(scoresActuels, lancer){ 
		let penalite = penalitePourAdversairesOuverts(scoresActuels, lancer);
		
		return scoresActuels.map(s => {
			return scorer(s, lancer, penalite)
		});
	}
}

const LIMITE = 3;

function penalitePourAdversairesOuverts(scores, lancer) {
	let scoreDuLanceur = trouverLeScoreDuLanceur(scores, lancer.lanceur);
	return calculerLaPenalitePourAdversairesOuverts(scoreDuLanceur.cible, lancer);
}

function trouverLeScoreDuLanceur(scores, lanceur) {
	return scores.filter(s => { return s.joueur === lanceur })[0];
}

function calculerLaPenalitePourAdversairesOuverts(cibleDuLanceur, lancer) {
	let impactsExistants = cibleDuLanceur[lancer.chiffre].impacts;
	let chiffreFermeApresImpacts = impactsExistants + lancer.impacts >= LIMITE;
	let surplusDimpacts = (impactsExistants + lancer.impacts) - LIMITE;
	
	return chiffreFermeApresImpacts ? surplusDimpacts * lancer.chiffre : 0;
}

function scorer(score, lancer, penalite) {
	return score.joueur === lancer.lanceur
		? scorerLeLanceur(score, lancer.chiffre, lancer.impacts)
		: scorerUnAdversaire(score, lancer.chiffre, penalite);
}

function scorerLeLanceur(scoreActuel, chiffre, nombreDimpacts) {
	let cibleDuLanceur = scoreActuel.cible;
	
	let nouveauxImpacts = impactsApresLeLancer(scoreActuel.cible, chiffre, nombreDimpacts);
	let nouvelleCible = impacterLaCible(cibleDuLanceur, chiffre, nouveauxImpacts);
	
	let nouveauScore = Object.assign({}, scoreActuel, {
		cible: nouvelleCible
	});
	
	return nouveauScore;
}

function impactsApresLeLancer(cible, chiffre, impactsDuLancer) {
	let impactsExistants = cible[chiffre].impacts;
	let nouveauxImpacts = Math.min(impactsExistants + impactsDuLancer, LIMITE);
	return nouveauxImpacts;
}

function impacterLaCible(cible, chiffreTouche, impacts) {
	return Object.assign({}, cible, {
		[chiffreTouche] : {
			impacts: impacts,
			ferme: impacts >= LIMITE
		}
	});
}

function scorerUnAdversaire(score, chiffre, nombreDePoints) {
	let nouveauxPoints = score.cible[chiffre].ferme
			? score.points
			: score.points + nombreDePoints;
	
	return Object.assign({}, score, {
		points: nouveauxPoints
	});
}
