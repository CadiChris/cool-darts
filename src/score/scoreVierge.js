'use striict';

export default function (joueur) {
	return {
		joueur: joueur,
		points: 0,
		cible: {
			15: { touches: 0, ferme: false},
			16: { touches: 0, ferme: false},
			17: { touches: 0, ferme: false},
			18: { touches: 0, ferme: false},
			19: { touches: 0, ferme: false},
			20: { touches: 0, ferme: false},
			25: { touches: 0, ferme: false}
		}
	};
}