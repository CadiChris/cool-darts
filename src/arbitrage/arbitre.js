'use strict'

import {toucher, penaliser, chiffreEstFerme} from "./score";

export function calculerLeNouveauScore(tableauDesScores, lancer) {
  let penalite = penaliteDuLancer(tableauDesScores, lancer)

  return tableauDesScores.map(s => {
    return scorer(s, lancer, penalite)
  });
}

const LIMITE = 3

function penaliteDuLancer(scores, lancer) {
  let scoreDuLanceur = trouverLeScoreDuLanceur(scores, lancer.lanceur)
  return calculerLaPenalite(scoreDuLanceur.cible, lancer)
}

function trouverLeScoreDuLanceur(scores, lanceur) {
  return scores.filter(s => {
    return s.joueur === lanceur
  })[0]
}

function calculerLaPenalite(cibleDuLanceur, lancer) {
  let touches = cibleDuLanceur[lancer.chiffre].touches
  let leChiffreVaSeFermer = touches + lancer.touches >= LIMITE
  let surplus = (touches + lancer.touches) - LIMITE

  return leChiffreVaSeFermer ? surplus * lancer.chiffre : 0
}

function scorer(score, lancer, penalite) {
  return score.joueur === lancer.lanceur
    ? toucher(lancer.chiffre, lancer.touches, score)
    : scorerUnAdversaire(score, lancer.chiffre, penalite)
}

function scorerUnAdversaire(score, chiffre, pointsDePenalite) {
  return chiffreEstFerme(chiffre, score)
    ? score
    : penaliser(pointsDePenalite, score)
}
