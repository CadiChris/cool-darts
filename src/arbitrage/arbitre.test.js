'use strict'
import assert from "assert";
import freeze from "deep-freeze";
import {scoreVierge} from "./score";
import * as arbitre from "./arbitre";

//	scores: [
//		{
//			joueur : 'J1',
//			cible : {
//				20 : {touches: 1|2|3, ferme:true|false},
//				19 : {},
//				...
//			},
//			points : 32
//		}
//		,
//		{
//			joueur: 'J2',
//			cible: {},
//			points: 0
//		}
//	]

const _0_TOUCHE = 0
const _1_TOUCHE = 1
const _2_TOUCHES = 2
const _3_TOUCHES = 3

describe('#calculerLeNouveauScore', () => {
  it('Ferme un chiffre après 3 touches', () => {
    let nouveauxScores = arbitre.calculerLeNouveauScore(
      [
        score('J1', {20: _2_TOUCHES}, 0)
      ],
      lancerDansLe(20, 'J1'))

    expect(nouveauxScores[0].cible).toMatchSnapshot()
  })

  it("N'augmente pas les touches d'un chiffre fermé", () => {
    let nouveauxScores = arbitre.calculerLeNouveauScore(
      [
        score('J1', {20: _3_TOUCHES}, 0)
      ],
      lancerDansLe(20, 'J1'))

    expect(nouveauxScores[0].cible).toMatchSnapshot()
  })

  it('Ne pénalise pas les adversaires si le lancer ne ferme pas le chiffre', () => {
    let nouveauxScores = arbitre.calculerLeNouveauScore(
      [
        score('J1', {20: _0_TOUCHE}, 0),
        score('J2', {20: _0_TOUCHE}, 0)
      ],
      lancerDansLe(20, 'J1'));

    expect(nouveauxScores).toMatchSnapshot()
  })

  it('Marque des points aux adversaires ouverts', () => {
    let nouveauxScores = arbitre.calculerLeNouveauScore(
      [
        score('J1', {20: _3_TOUCHES}, 0),
        score('J2', {20: _0_TOUCHE}, 10)
      ],
      lancerDansLe(20, 'J1'));

    assert.strictEqual(nouveauxScores[1].points, 30);
  });

  it('Ne marque pas de points aux adversaires fermés', () => {
    let nouveauxScores = arbitre.calculerLeNouveauScore(
      [
        score('J1', {20: _3_TOUCHES}, 0),
        score('J2', {20: _3_TOUCHES}, 0)
      ],
      lancerDansLe(20, 'J1'));

    assert.strictEqual(nouveauxScores[1].points, 0);
  });

  it("Préserve l'ordre des scores", () => {
    let nouveauxScores = arbitre.calculerLeNouveauScore(
      [
        score('J1', {20: _0_TOUCHE}, 0),
        score('J2', {20: _0_TOUCHE}, 0),
        score('J3', {20: _0_TOUCHE}, 0)
      ],
      lancerDansLe(20, 'J2'));

    assert.strictEqual(nouveauxScores[0].joueur, 'J1');
    assert.strictEqual(nouveauxScores[1].joueur, 'J2');
    assert.strictEqual(nouveauxScores[2].joueur, 'J3');
  })
})

function score(leJoueur, sesChiffres, sesPoints) {
  let score = Object.assign({}, scoreVierge(leJoueur), {
    points: sesPoints,
    cible: cible(sesChiffres)
  });
  freeze(score);
  return score;
}

function cible(chiffres) {
  let cible = scoreVierge().cible;
  for (const c in chiffres) {
    cible[c] = {
      touches: chiffres[c],
      ferme: chiffres[c] === _3_TOUCHES
    };
  }
  freeze(cible);
  return cible;
}

function lancerDansLe(chiffre, lanceur, touches = 1) {
  let lancer = {
    lanceur: lanceur,
    chiffre: chiffre,
    touches: touches
  };
  freeze(lancer);
  return lancer;
}