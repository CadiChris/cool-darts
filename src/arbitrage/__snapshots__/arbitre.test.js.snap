exports[`#calculerLeNouveauScore Ferme un chiffre après 3 touches 1`] = `
Object {
  "15": Object {
    "ferme": false,
    "touches": 0,
  },
  "16": Object {
    "ferme": false,
    "touches": 0,
  },
  "17": Object {
    "ferme": false,
    "touches": 0,
  },
  "18": Object {
    "ferme": false,
    "touches": 0,
  },
  "19": Object {
    "ferme": false,
    "touches": 0,
  },
  "20": Object {
    "ferme": true,
    "touches": 3,
  },
  "25": Object {
    "ferme": false,
    "touches": 0,
  },
}
`;

exports[`#calculerLeNouveauScore N'augmente pas les touches d'un chiffre fermé 1`] = `
Object {
  "15": Object {
    "ferme": false,
    "touches": 0,
  },
  "16": Object {
    "ferme": false,
    "touches": 0,
  },
  "17": Object {
    "ferme": false,
    "touches": 0,
  },
  "18": Object {
    "ferme": false,
    "touches": 0,
  },
  "19": Object {
    "ferme": false,
    "touches": 0,
  },
  "20": Object {
    "ferme": true,
    "touches": 3,
  },
  "25": Object {
    "ferme": false,
    "touches": 0,
  },
}
`;

exports[`#calculerLeNouveauScore Ne pénalise pas les adversaires si le lancer ne ferme pas le chiffre 1`] = `
Array [
  Object {
    "cible": Object {
      "15": Object {
        "ferme": false,
        "touches": 0,
      },
      "16": Object {
        "ferme": false,
        "touches": 0,
      },
      "17": Object {
        "ferme": false,
        "touches": 0,
      },
      "18": Object {
        "ferme": false,
        "touches": 0,
      },
      "19": Object {
        "ferme": false,
        "touches": 0,
      },
      "20": Object {
        "ferme": false,
        "touches": 1,
      },
      "25": Object {
        "ferme": false,
        "touches": 0,
      },
    },
    "joueur": "J1",
    "points": 0,
  },
  Object {
    "cible": Object {
      "15": Object {
        "ferme": false,
        "touches": 0,
      },
      "16": Object {
        "ferme": false,
        "touches": 0,
      },
      "17": Object {
        "ferme": false,
        "touches": 0,
      },
      "18": Object {
        "ferme": false,
        "touches": 0,
      },
      "19": Object {
        "ferme": false,
        "touches": 0,
      },
      "20": Object {
        "ferme": false,
        "touches": 0,
      },
      "25": Object {
        "ferme": false,
        "touches": 0,
      },
    },
    "joueur": "J2",
    "points": 0,
  },
]
`;
